core/arch/aarch64/build_ldstex.c

```diff
@@ -0,0 +1,297 @@
+/* **********************************************************
+ * Copyright (c) 2017 ARM Limited. All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of ARM Limited nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* A wrapper round the decoder that recognises single-entry single-exit blocks
+ * of contiguous instructions containing an exclusive load/store pair and bundles
+ * them into a macro-instruction, OP_ldstex. This is a temporary solution for
+ * i#1698 and is likely to be fragile. Known problems:
+ *
+ * - We only handle single-entry single-exit contiguous code blocks. (Usually
+ *   they are written as inline assembler so they do fit this pattern.)
+ * - If the block uses all of X0-X5 and the stolen register then we cannot
+ *   mangle it (so it is better not to recognise it at all).
+ * - The contents of an OP_ldstex cannot be instrumented.
+ * - If execution remains in an OP_ldstex then signal delivery may be delayed.
+ * - Bad things might happen if there is a SIGSEGV or SIGBUS in an OP_ldstex.
+ * - Code flushing.
+ *
```

Comment:
Missing from the list is a major problem: this does not support instrumentation between the load and store which makes it a non-starter for taint tracking and other tools.